### PR TITLE
Standardize visualization signatures and styling.

### PR DESCRIPTION
**Pull request by nonhermitian**
### Summary
An overhaul of the counts and states visualization functions (both mpl and interactive) to standardize the function signatures, styling, and allow for more functionality.

### Details and comments
- The signatures of all functions are now as close to identical as possible, and follow the standard mpl way to doing things. 
- The styling has also been updated. 
- It is now possible to change the figure size in every function, and it is possible to change the colors used in most functions (e.g. it is now possible to change a single bar color in `plot_histogram`)
- The interactive and mpl functions are no longer called together.  Instead the interactive ones must be explicitly called using the 'i'+plot_routine naming convention.
- Each mpl based visualization is loaded regardless of whether mpl is installed or not, but gives an `ImportError: Matploltib not installed.` if trying to call the function and mpl is not installed.
- The city plots are now in the same row rather than same column.

### CODE REVIEW

**Comment by jaygambetta** [comment]
This looks good (after test passing) but can @ajavadia and @mtreinish could you look as well so that we can converge on this interface being defined.

**Comment by mtreinish** [comment]
So from a quick glance (I still need to go through this in depth and test all the methods locally) I'm a bit confused as to what we're defining as our public interface as for the state visualizations now. By adding all the individual functions for the different plot types is the intent there to make each one of those functions part of our public interface? Before, we just had the `plot_state()` function and the different types of plots specified by the `method` kwarg. That function still exists here and I wonder why we need both as part of the interface.

My only concern is that the `plot_state()` function is the only place that does the data checks to make sure it's a square matrix and make sure it's a multiqubit state. It also still converts a statevector to density matrix too. Not that any of these operations or checks are particularly difficult it's just a usability thing. If we say they work the same way and our part of our interface, but then I give a malformed matrix to the individual methods it'll try to plot it but for plot_state() it will error saying this isn't a square matrix or whatever. Ditto for a statevector input, although that probably doesn't matter because we say in the docstring that rho is an array of dimensions 2**n x 2**nn complex numbers.

**Comment by nonhermitian** [comment]
I would say have separate functions because they do not all have same structure to their arguments. I should do an array shape check though in all of them, and do outer produt when needed.

**Comment by mtreinish** [comment]
So I started running through my little test scripts that I put together when I was working on #1229 and the I hit an issue on `plot_bloch_vector()` when it's called via `plot_state()` (which does the loop over the qubits).

```
Traceback (most recent call last):
  File "test_hinton.py", line 37, in <module>
    fig = visualization.plot_state(rho_ghz, method='bloch')
  File "/home/mtreinish/git/qiskit/qiskit-core/qiskit/tools/visualization/_state_visualization.py", line 545, in plot_state
    figsize=figsize)
  File "/home/mtreinish/git/qiskit/qiskit-core/qiskit/tools/visualization/_state_visualization.py", line 145, in plot_bloch_vector
    fig.set_size_inches(figsize[0], figsize[1])
AttributeError: 'NoneType' object has no attribute 'set_size_inches'
```

It looks like this is being caused because when we set ax in the Bloch object it doesn't have an internal figure object and relies on the one we created externally. So `B.fig` is `None` in that case.

**Comment by nonhermitian** [comment]
@mtreinish please take another look.  

- Can now pass state vector or density matrix to state visualizers.
- fixed the issue when `Bloch` is passed an axes instance.
- `plot_state` is set to be depreciated in two releases (0.9).

File: qiskit/tools/visualization/_counts_visualization.py [hunk 1]
```diff
@@ -70,13 +63,14 @@ def plot_histogram(data, number_to_keep=None, legend=None, options=None,
...
-        _, ax = plt.subplots()
-
+    fig, ax = plt.subplots(figsize=figsize)
     labels = sorted(
         functools.reduce(lambda x, y: x.union(y.keys()), data, set()))
+
+    # Set bar colors
+    if color is None:
+        color = ['#648fff', '#dc267f', '#785ef0', '#ffb000', '#fe6100']
+
```
[the remark this answers is not in the document]
Comment (nonhermitian) [reply qiskit/tools/visualization/_counts_visualization.py]: Indeed that was my error.  It is fixed.

File: qiskit/tools/visualization/_counts_visualization.py [hunk 2]
```diff
@@ -14,54 +14,47 @@
...
-                   filename=None, show=False):
+
+def plot_histogram(data, figsize=(7, 5), color=None, number_to_keep=None,
+                   sort='asc', legend=None, bar_labels=True):
     """Plot a histogram of data.
 
     Args:
         data (list or dict): This is either a list of dictionaries or a single
             dict containing the values to represent (ex {'001': 130})
-        number_to_keep (int): DEPRECATED the number of terms to plot and rest
```
[the remark this answers is not in the document]
Comment (nonhermitian) [reply qiskit/tools/visualization/_counts_visualization.py]: Yep, forgot to remove.

File: qiskit/tools/visualization/_counts_visualization.py [hunk 3]
```diff
@@ -14,54 +14,47 @@
...
     """
-    if options is None:
-        options = {}
-
+    if not HAS_MATPLOTLIB:
+        raise ImportError('Must have Matplotlib installed.')
     if number_to_keep is not None:
         warnings.warn("number_to_keep has been deprecated, use the options "
                       "dictionary and set a number_to_keep key instead",
                       DeprecationWarning)
```
[the remark this answers is not in the document]
Comment (nonhermitian) [reply qiskit/tools/visualization/_counts_visualization.py]: yes.

**Comment by jaygambetta** [comment]
if we are going to depreciate plot_state please make sure the tutorial explains all the plotting methods clearly.

**Comment by mtreinish** [comment]
@jaygambetta yeah, I'm planning to push an update to the notebook after this merges

File: qiskit/tools/visualization/_state_visualization.py [hunk 4]
```diff
@@ -420,56 +511,46 @@ def plot_state_qsphere(rho, filename=None, show=False):
...
+        ImportError: Requires matplotlib.
         VisualizationError: if the input is not a statevector or density
         matrix, or if the state is not an multi-qubit quantum state.
     """
-
+    if not HAS_MATPLOTLIB:
+        raise ImportError('Must have Matplotlib installed.')
+    warnings.warn("'plot_state' will be depreciated in terra-0.9.  \
+                  Call the desired plotting routine directly.",
+                  DeprecationWarning)
```
Comment (mtreinish) [inline qiskit/tools/visualization/_state_visualization.py]: I would slightly reword this to say something like:

plot_state() is deprecated and will be removed in the 0.9 release.  Instead you can use the plot_state_*() function directly for method being used.

File: qiskit/tools/visualization/_state_visualization.py [hunk 5]
```diff
@@ -5,59 +5,89 @@
...
+    if len(shape) != 2 or shape[0] != shape[1]:
+        raise VisualizationError("Input is not a valid quantum state.")
+    # Check state is an n-qubit state
+    num = int(np.log2(len(rho)))
+    if 2 ** num != len(rho):
+        raise VisualizationError("Input is not a multi-qubit quantum state.")
+    return rho
+
+
+def plot_hinton(rho, title='', figsize=None):
```
Comment (mtreinish) [inline qiskit/tools/visualization/_state_visualization.py]: We probably should update the function name here to be `plot_state_hinton()` to be consistent with the other functions

File: qiskit/tools/visualization/_state_visualization.py [hunk 6]
```diff
@@ -99,57 +127,77 @@ def plot_hinton(rho, title='', filename=None, show=False):
...
+        ax2.set_xticks(np.arange(0, lx+0.5, 1))
+        ax2.set_yticks(np.arange(0, ly+0.5, 1))
+        ax2.set_yticklabels(row_names, fontsize=14)
+        ax2.set_xticklabels(column_names, fontsize=14, rotation=90)
     ax2.autoscale_view()
     ax2.invert_yaxis()
-    ax2.set_title('Imag[rho]')
+    ax2.set_title('Imag[rho]', fontsize=14)
     if title:
         plt.title(title)
```
Comment (mtreinish) [inline qiskit/tools/visualization/_state_visualization.py]: This is an existing bug, but when I called:

plot_hinton(rho, title='My Hinton') this sets the title over ax2 instead of Imag[rho]

File: qiskit/tools/visualization/_state_visualization.py [hunk 7]
```diff
@@ -173,79 +221,105 @@ def plot_state_city(rho, title="", filename=None, show=False):
...
+    ax2.axes.set_zlim3d(np.min(dzi), np.max(dzi)+eps)
+    ax2.w_xaxis.set_ticklabels(row_names, fontsize=14, rotation=45)
+    ax2.w_yaxis.set_ticklabels(column_names, fontsize=14, rotation=-22.5)
     # ax2.set_xlabel('basis state', fontsize=12)
     # ax2.set_ylabel('basis state', fontsize=12)
-    ax2.set_zlabel("Imag[rho]")
+    ax2.set_zlabel("Imag[rho]", fontsize=14)
+    for tick in ax2.zaxis.get_major_ticks():
+        tick.label.set_fontsize(14)
     plt.title(title)
```
Comment (mtreinish) [inline qiskit/tools/visualization/_state_visualization.py]: Just like with hinton this is only setting the title over the ax2 plot (it's also a very small fontsize)

File: qiskit/tools/visualization/_state_visualization.py [hunk 8]
```diff
@@ -420,56 +511,46 @@ def plot_state_qsphere(rho, filename=None, show=False):
...
     if method == 'city':
-        fig = plot_state_city(rho, filename=filename, show=show)
+        fig = plot_state_city(rho, figsize=figsize)
     elif method == "paulivec":
-        fig = plot_state_paulivec(rho, filename=filename, show=show)
+        fig = plot_state_paulivec(rho, figsize=figsize)
     elif method == "qsphere":
-        fig = plot_state_qsphere(rho, filename=filename, show=show)
+        fig = plot_state_qsphere(rho, figsize=figsize)
     elif method == "bloch":
```
Comment (mtreinish) [inline qiskit/tools/visualization/_state_visualization.py]: Should we add a new function for the multiqubit bloch vector? Right now when we remove plot_state() we'll lose this functionality.

File: qiskit/tools/visualization/_state_visualization.py [hunk 9]
```diff
@@ -321,8 +395,25 @@ def phase_to_color_wheel(complex_number):
     return color_map[angle_round]
 
 
-def plot_state_qsphere(rho, filename=None, show=False):
-    """Plot the qsphere representation of a quantum state."""
+def plot_state_qsphere(rho, figsize=None):
```
Comment (mtreinish) [inline qiskit/tools/visualization/_state_visualization.py]: When running this we print the eigenvalues before generating the visualization. Do you think that is still needed? It's a bit weird to have the plot function print to stdout

File: qiskit/tools/visualization/_state_visualization.py [hunk 10]
```diff
@@ -173,79 +221,105 @@ def plot_state_city(rho, title="", filename=None, show=False):
...
     ax.set_yticks([-1, -0.5, 0, 0.5, 1])
-    ax.set_xticklabels(labels, fontsize=12, rotation=70)
-    ax.set_xlabel('Pauli', fontsize=12)
+    ax.set_xticklabels(labels, fontsize=14, rotation=70)
+    ax.set_xlabel('Pauli', fontsize=14)
     ax.set_ylim([-1, 1])
+    ax.set_facecolor('#eeeeee')
+    for tick in ax.xaxis.get_major_ticks()+ax.yaxis.get_major_ticks():
+        tick.label.set_fontsize(14)
     plt.title(title)
```
Comment (mtreinish) [inline qiskit/tools/visualization/_state_visualization.py]: When I set a title it was a pretty small font, I wonder if we should change the font to make it larger or at least bold

File: qiskit/tools/visualization/_state_visualization.py [hunk 11]
```diff
@@ -482,23 +563,20 @@ def plot_state(quantum_state, method='city', filename=None, show=False):
...
-            plot_bloch_vector(bloch_state, "qubit " + str(i), ax=ax)
-        if filename:
-            plt.savefig(filename)
-        elif show:
-            plt.show()
+            plot_bloch_vector(bloch_state, "qubit " + str(i), ax=ax,
+                              figsize=figsize)
     elif method == "wigner":
-        fig = plot_wigner_function(rho, filename=filename, show=show)
+        fig = plot_wigner_function(rho)
```
Comment (mtreinish) [inline qiskit/tools/visualization/_state_visualization.py]: We don't expose this function via `qiskit.tools.visualization.__init__` so we'll drop the functionality when we remove `plot_state()` I'm fine with doing that since this method always felt a bit out of place, and given we need to come to a resolution in #1106 and #1302 I think if we want to make wigner part of the public api we do it as a separate thing.

File: qiskit/tools/visualization/interactive/_iplot_state.py [hunk 12]
```diff
@@ -18,14 +18,14 @@
 from ._iplot_qsphere import iplot_qsphere
 
 
-def iplot_state(quantum_state, method='city', options=None):
+def iplot_state(quantum_state, method='city', figsize=None):
```
Comment (mtreinish) [inline qiskit/tools/visualization/interactive/_iplot_state.py]: Should we also be raising a deprecation warning here? There are more kwargs for the individual method functions so it feels like we'll want to to do the same thing here that we did in the non-interactive version. Ditto for the input data check (maybe move `_validate_input_state` to `qiskit.tools.visualization._utils` and use it for the interactive functions too.

File: qiskit/tools/visualization/_state_visualization.py (hunk 4 above)
Comment (nonhermitian) [reply qiskit/tools/visualization/_state_visualization.py]: The `plot_bloch_vector` does not follow this pattern, nor does `plot_wigner_function`  Rather than have a lengthy depreciation warning, the user can turn to the tutorials or docs for help.

File: qiskit/tools/visualization/_state_visualization.py (hunk 5 above)
Comment (nonhermitian) [reply qiskit/tools/visualization/_state_visualization.py]: done.

File: qiskit/tools/visualization/_state_visualization.py (hunk 4 above)
Comment (mtreinish) [reply qiskit/tools/visualization/_state_visualization.py]: Well plot_bloch_vector is a different function that was already part of the api, and has different inputs. That was actually a comment further down about how we lose the functionality to go from rho to multiple bloch vectors after we remove `plot_state()`. As for plot_wigner_function() I'm actually thinking more that we should remove it or at least keep it separate from plot_state() as a solution to #1106 and #1302 .

Part of my issue with the wording here is that it assumes people know what their desired plotting routine is, which they may not if they've only ever used `plot_state()` before. Especially since we don't have anything in the changelog about this yet. If ~50 more characters is too long:

plot_state is deprecated and will be removed in the 0.9 release. Use the plot_state_* functions instead.

Which is only an additional 5.

**Comment by nonhermitian** [comment]
I believe that all known issues are resolved.  If this looks good, I will go ahead and update the changelog.

**Comment by mtreinish** [comment]
@nonhermitian yeah I think this looks good, so lets update the changelog and then I'll approve.
